Update RGMB block name and region ID assignment to catch potential ID assignment conflicts

shikhar413 opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->
refs #21120

shikhar413 committed:
Remove Cartesian pin rotation block, add logic to copy sides from outer_core to sideset 10000

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1054319/) on 6b35e6e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21121/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21121/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e0d954c27a3c9d21fc36765fe784fe7ca6cc8a7f`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1054650/) on ba409b0 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21121/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1055184/) on 141c332 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>28ac7e</th>
      <th colspan=4>#21121 141c33</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.58%</td>
      <td>82.58%</td>
      <td>+0.00%</td>
      <td>87.50%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>75293</td>
      <td>75307</td>
      <td>+14</td>
      <td>14</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15888</td>
      <td>15890</td>
      <td>+2</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21121/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21121/coverage/framework)


## Modules coverage

### Reactor

<table>
  <thead>
    <tr>
      <th></th>
      <th>28ac7e</th>
      <th colspan=4>#21121 141c33</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.28%</td>
      <td>93.85%</td>
      <td>-1.43%</td>
      <td>80.84%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3087</td>
      <td>3143</td>
      <td>+56</td>
      <td>270</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>153</td>
      <td>206</td>
      <td>+53</td>
      <td>64</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21121/coverage_diff/reactor)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21121/coverage/reactor)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21121/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 87.50% is less than the suggested 90.0%
- `reactor` new line coverage rate 80.84% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1056850/) on 5ca2ab2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21121/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21121/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format abb9fa1c8ab66cf64ab88f9572dc1a14d66ee5ea`

GiudGiud commented on framework/include/meshgenerators/MeshGenerator.h:
```diff
@@ -203,6 +212,29 @@ MeshGenerator::declareMeshProperty(const std::string & data_name, const T & init
+
+  T & data = restartable_data_ref.set();
+  data = init_value;
+
```
also isnt there concerns if the mesh generator has already be ran?

shikhar413 replied:
For some context, this function basically allows you to overwrite the value of `data_name` from another mesh generator. I wasn't sure which checks I should add to allow such behavior but this function would basically be used to update the values of parameters of mesh generators that have already been called from an external mesh generator, since to my knowledge this functionality is not available in MOOSE MG's and we need a way to keep track of block id's that have already been assigned at each stage of PinMeshGenerator, AssemblyMeshGenerator, and CoreMeshGenerator. This block ID mapping is being stored by an intermediate MG ReactorMeshParams, and each of PMG, AMG, and CMG will query ReactorMeshParams for the block ID mapping, then update this block ID mapping after their respective constructors and `generate` functions have been called.

shikhar413 replied:
This is now removed

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -150,6 +150,19 @@ PatternedMeshGenerator::generate()
       // Move the mesh into the right spot.  -i because we are starting at the top
       MeshTools::Modification::translate(cell_mesh, deltax, -deltay, 0);
 
+      // Define a reference map variable for subdomain map
+      auto & main_subdomain_map = _row_meshes[i]->set_subdomain_name_map();
+      // Retrieve subdomain name map from the mesh to be stitched and insert it to the main
+      // subdomain map
+      const auto & increment_subdomain_map = cell_mesh.get_subdomain_name_map();
+      main_subdomain_map.insert(increment_subdomain_map.begin(), increment_subdomain_map.end());
+      // Check if one SubdomainName is shared by more than one subdomain ids
+      std::set<SubdomainName> main_subdomain_map_name_list;
```
it's good to comment every line but what's the goal here?
The check on subdomain ids?

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -163,8 +176,23 @@ PatternedMeshGenerator::generate()
   // We're going to stitch them all to row 0 (which is the real mesh)
   for (MooseIndex(_pattern) i = 1; i < _pattern.size(); i++)
+  {
+    // Define a reference map variable for subdomain map
+    auto & main_subdomain_map = _row_meshes[0]->set_subdomain_name_map();
+    // Retrieve subdomain name map from the mesh to be stitched and insert it to the main
+    // subdomain map
+    const auto & increment_subdomain_map = _row_meshes[i]->get_subdomain_name_map();
+    main_subdomain_map.insert(increment_subdomain_map.begin(), increment_subdomain_map.end());
+    // Check if one SubdomainName is shared by more than one subdomain ids
+    std::set<SubdomainName> main_subdomain_map_name_list;
+    for (auto const & id_name_pair : main_subdomain_map)
+      main_subdomain_map_name_list.emplace(id_name_pair.second);
+    if (main_subdomain_map.size() != main_subdomain_map_name_list.size())
+      paramError("inputs", "The input meshes contain subdomain name maps with conflicts.");
+
```
this is the same code right?
Should this be a routine then?

shikhar413 replied:
Created the function `mergeSubdomainNameMaps`

GiudGiud commented on modules/reactor/include/meshgenerators/ReactorGeometryMeshBuilderBase.h:
```diff
@@ -0,0 +1,87 @@
+  ///The ReactorMeshParams object that is storing the reactor global information for this reactor geometry mesh
+  MeshGeneratorName _reactor_params;
+
+  ///Tracker to assign new block ids in RGMB mesh generators.
+  ///Increments by 1 each time a new block id is requested
+  subdomain_id_type _current_block_id;
```
@roystgnr isnt this already the default behavior by libmesh?

roystgnr replied:
Sorry, I missed this conversation originally somehow.

There's no mechanism in libMesh to request a subdomain id; the best you can do is what I mentioned in another thread, `get_mesh_subdomains()` on a prepared mesh, then take the `back()` of the resulting `set` and add 1.

shikhar413 replied:
I have restructured the code now so that the next free id is retrieved from the max(`get_mesh_subdomains`) + 1

GiudGiud reviewed (commented):
quick comments, have not gotten to the core of this

GiudGiud commented on modules/reactor/src/meshgenerators/AssemblyMeshGenerator.C:
```diff
@@ -384,121 +420,124 @@ AssemblyMeshGenerator::AssemblyMeshGenerator(const InputParameters & parameters)
 std::unique_ptr<MeshBase>
 AssemblyMeshGenerator::generate()
 {
+  // Re-initialize name_id_map and current_block_id from ReactorMeshParams object
+  // in case variables have been modified between constructor and this method
+  initializeReactorMeshParams();
```
this is not really intuitive.
We need some sort of hasChanged() api similar to is_prepared() for the meshes.

GiudGiud replied:
well it s mostly for the manual block id / name tracking so I guess it might be removed

shikhar413 replied:
This has been removed now

GiudGiud commented on modules/reactor/src/meshgenerators/AssemblyMeshGenerator.C:
```diff
@@ -384,121 +420,124 @@ AssemblyMeshGenerator::AssemblyMeshGenerator(const InputParameters & parameters)
-    mooseError("Expected mesh inputs to have extra integer ids");
+    mooseError("Expected mesh inputs to have region_id and pin_type_id extra integer ids");
 
   unsigned int ptid_int = (*_build_mesh)->get_elem_integer_index(pin_type_id_name);
   unsigned int rid_int = (*_build_mesh)->get_elem_integer_index(region_id_name);
```
ptid_int, rid_int etc 
are all unintuitive names

shikhar413 replied:
Renamed

GiudGiud commented on modules/reactor/src/meshgenerators/AssemblyMeshGenerator.C:
```diff
@@ -384,121 +420,124 @@ AssemblyMeshGenerator::AssemblyMeshGenerator(const InputParameters & parameters)
   std::string plane_id_name = "plane_id";
   std::string region_id_name = "region_id";
   std::string pin_type_id_name = "pin_type_id";
+  std::string assembly_type_id_name = "assembly_type_id";
 
   if (!(*_build_mesh)->has_elem_integer(region_id_name) ||
       !(*_build_mesh)->has_elem_integer(pin_type_id_name))
-    mooseError("Expected mesh inputs to have extra integer ids");
+    mooseError("Expected mesh inputs to have region_id and pin_type_id extra integer ids");
 
   unsigned int ptid_int = (*_build_mesh)->get_elem_integer_index(pin_type_id_name);
   unsigned int rid_int = (*_build_mesh)->get_elem_integer_index(region_id_name);
 
-  // go through the background and duct regions and assign their elements
-  // pin_type id placeholders and region id place holders
-  for (const auto i : make_range(_peripheral_regions))
-  {
-    subdomain_id_type region = UINT16_MAX - (i + 1);
-
-    for (const auto & elem : as_range((*_build_mesh)->active_subdomain_elements_begin(region),
-                                      (*_build_mesh)->active_subdomain_elements_end(region)))
-    {
-      elem->set_extra_integer(ptid_int, region); // identifier for element not in a pin
-      elem->set_extra_integer(rid_int, region);
-    }
-  }
+  unsigned int atid_int, pid_int = 0;
+  if (!(*_build_mesh)->has_elem_integer(assembly_type_id_name))
+    atid_int = (*_build_mesh)->add_elem_integer(assembly_type_id_name);
+  else
+    atid_int = (*_build_mesh)->get_elem_integer_index(assembly_type_id_name);
 
   if (_extrude)
   {
-    // swap the region ids on the subdomain ids to the correct ones
-    // for their axial layer
     if (!(*_build_mesh)->has_elem_integer(plane_id_name))
-      mooseError("Expected extruded mesh to have plane IDs");
+      mooseError("Expected extruded mesh to have plane_id extra integers");
+    pid_int = (*_build_mesh)->get_elem_integer_index(plane_id_name);
```
you do this thing 3 times, let's have a routine

shikhar413 replied:
Created the function `getElemIntegerFromMesh`

GiudGiud commented on modules/reactor/src/meshgenerators/ReactorGeometryMeshBuilderBase.C:
```diff
@@ -0,0 +1,95 @@
+  this->declareMeshProperty(_reactor_params, "name_id_map", _name_id_map);
+}
+
+dof_id_type
+ReactorGeometryMeshBuilderBase::getBlockId(const std::string block_name,
+                                           const dof_id_type region_id)
+{
+  if (_name_id_map.find(block_name) != _name_id_map.end())
+  {
+    // Block name exists in map, verify region_id matches one in map and return block_id
+    if (_name_id_map[block_name].second != region_id)
+      mooseError("Provided region id ",
+                 region_id,
+                 " for block name ",
+                 block_name,
+                 " does not match the one that was provided in an earlier mapping - ",
+                 _name_id_map[block_name].second);
+    return _name_id_map[block_name].first;
+  }
+  else
+  {
+    // Block name does not exist, return a new block id and store in _name_id_map
+    const auto new_block_id = _current_block_id++;
+    _name_id_map[block_name] = std::make_pair(new_block_id, region_id);
+    return new_block_id;
+  }
```
@roystgnr 
they are adding this manual tracking of blocks within this MG.
Is this possible in libmesh already? @shikhar413 and I looked but didnt find. 

If not then I think it belongs there. What do you think?

roystgnr replied:
I'm not sure what the code here is trying to do.  LibMesh gives you a 1-to-1 mapping between subdomain names and subdomain ids (`MeshBase::subdomain_name()` and the how-did-I-let-Cody-slip-this-API-name-past `MeshBase::get_id_by_name()`), but we have no concept of "regions".

GiudGiud replied:
Ok I’ll have another look. Thanks for pitching in

shikhar413 replied:
For more context, these functions basically check that the subdomain name is mapped uniquely to (subdomain_id, region_id), where region_id is an extra element integer. For simplicity, assume that both subdomain name and region id is provided by the user. However, we would need to generate a _unique_ new subdomain id if it doens't already exist in the mesh. Is there a libmesh function that will allow us to do this? In the interim, I am manually creating a workaround for this, where `_current_block_id` will increment by 1 if a new block id is requested, and then this block name - (block_id, region_id) mapping is stored in `_name_id_map`

roystgnr replied:
`MeshBase::get_mesh_subdomains()` gives you an ordered set of subdomain ids, and you can just pick the max of that plus 1 to get a new unique subdomain id, but the set is only updated on `prepare_for_use()` (or if you want to do it manually, just `cache_elem_data()` is enough) since it requires a loop over elements and a parallel `set_union()`.

GiudGiud replied:
so if we use what Roy suggests, what does this map provide? A way to check that region ids are set consistently across block IDs ?

I think this map might actually be counter productive, since it forces a unique mapping block name -> region id and the whole purpose of region id is not have this uniqueness. We want multiple region ids in a given block, because we can only have so many blocks before we stop scaling

shikhar413 replied:
@GiudGiud , there seems to be some confusion about the purpose of _name_id_map. The map is there just to check that the user has not tried to define a block_name - region_id pairing that is inconsistent with previously defined pairs in the other RGMB MG's. For this reason, we need this variable to persist throughout the duration of the RGMB mesh generation stage (from pin mesh generation all the way up to core mesh generation). However, notice that the key in the map is the block name, not the region id. Therefore, you can have multiple blocks that share the same region id. In fact, this is done in the code (and you can see this in some of the unit test cases as well). For triangular regions, the "_TRI" suffix is appended to the block name, but they will share the same region ID as the quad region, so you can have the following block name - region id pairings:

RGMB_BLOCK_1 : 1
RGMB_BLOCK_TRI : 1

In both cases, we need to make sure the two block names are associated with unique block ids, hence the discussion with Roy about how to generate block ids that do not cause any contention with previously defined MGs. The user should not have to worry about what these block id's are, they are used and tracked internally by the RGMB MG's and stored in the map as a block_name: (block_id, region_id) mapping.

One other thing to note is that all of these mappings should only be relevant for RGMB MG's, i.e. none of the other MG's need to have this design constraint enforced. These MG's were designed under the assumption that these MGs would be completely self-contained so that other MG's would not need to be used before, during, or after the RGMB MG stage, but if users choose to do so, then it is up to them that the other MGs are consistent with how RGMB has generated its mesh

GiudGiud replied:
> notice that the key in the map is the block name, not the region id

that is the problem. That means you can only have as many region ids as you have blocks. 
The key, if you wanted to do this, which we're more and more leaning against should be the region id because that is the one that you have plenty of.

shikhar413 replied:
I'm not sure I understand. How would you represent a block with a mixture of tri and quad elements? If the key were a unique region id among the two, then there'd be no way to represent both elements as having a shared region id since the unique key would have to point to either the quad elements or the tri elements.

Maybe there is a misunderstanding in terms of what region id represents? Here, region id is a single extra element id to represent like materials, like in Griffin. For example, if we have a fuel pin with tri and quad elements, these would have to be unique blocks, but they can be assigned the same region ID to indicate that these elements have the same material. All other extra elements integers (such as pin id, assembly id, plane id etc) are assigned on their own, independent of the region id, and therefore can be combined together for a single block.

> That means you can only have as many region ids as you have blocks.

I don't really see how the number of region ids could outnumber the number of blocks

GiudGiud replied:
The whole purpose of using region ids is to be able to be able to have multiple region ids within a block.
For example in an assembly you could have 1 block: fuel and 284 different region ids, which correspond to the cross section regions. Your second paragraph shows we agree on this purpose. This map precludes this, because you can only have less regions than blocks when they map to blocks with a block_name key. 

You do not represent a block with a mixture of tri and quad, you use two different blocks for this but then you use the same region id if appropriate. This is the exception case where you need one more block than region id. The general case is that we want more region ids than blocks. 

You can have a key of region id and a vector of blocks as the item. That is again if we assume we want this map.

Maybe we should meet. Available tomorrow after (Wednesday) or Thursday 2pm +.

shikhar413 replied:
I think I see what you're saying. I have to think about this more but with this new design, the need for this map isn't obvious to me anymore, since we no longer are imposing any constraints on blocks / region ids. The only additional functionality that I think we should keep is just having the ability to generate unique non-clashing block ids for every block name that isn't already in the mesh, which I think can be done entirely on the libMesh side without the use of this map.

Down for a call, tomorrow afternoon works for me

shikhar413 replied:
@roystgnr is there a way to call `cache_elem_data()` in the mesh generator constructor in order to figure out what the max block id so far is? These subdomain id values are set in the constructors of the mesh generators in this PR by calling `addMeshSubgenerator` for existing mesh generators, so this max value is needed before the mesh is actually generated in order to ensure that these subdomain ids don't potentially conflict with other MGs later on

shikhar413 replied:
I'm playing around with the `cache_elem_data()` and  `get_mesh_subdomains()` functions, it looks like it only returns the subdomain ids of mesh created within the scope of the current mesh generator block, is there a way to return subdomain ids of _all mesh generators up to the current mesh generator_ under consideration? Basically if there are two pin mesh generators that are called successively (and not necessarily in communication with each other), I would like to know the block ids used by the first pin mesh generator so that the second pin mesh generator can set its block ids without conflicting with those set by the first mesh generator.

roystgnr replied:
libMesh doesn't know what a mesh generator even is.  `cache_elem_data()` will calculate on whichever mesh you call it on, `get_mesh_subdomains()` will return data for whichever mesh you call it on; if you have multiple `MeshBase` objects being operated on then you'll need to make those calls on each (and merge the results of the get afterwards).  If your first pin mesh generator is an input to your second pin mesh generator then you've got an easy time of it; if not you're out of luck.

Your best bet in the latter case might be to handle any conflicts at the point of merging.  This is what libMesh does when stitching two meshes together, for instance; if one mesh has an "element 1" and the another has a different "element 1" then we don't just throw our hands up and say the meshes conflict, we renumber the latter's elements (calculating an offset to add to element ids, node ids, unique_ids) to solve the conflict.

shikhar413 replied:
Thanks for that explanation @roystgnr. That makes a lot of sense, unfortunately for us, the situation you described in the latter case is what is ultimately required for our purposes so some additional effort is definitely required.

@GiudGiud do you have any thoughts on this? Right now I am thinking that in the constructor we assign block id's as a temporary value (something like UINT_MAX - 1 - elem_index) in the constructor, and then try to merge all block name / id info and remap all element ids/names in the `generate` function later. The implementation won't be very straightforward or clean, however. It'd definitely be easier to let the user try to set the block id's with the block names and region ids instead. Also, given that a mesh generator has no idea what block id's have been assigned up to the point when it's been called, this also does not address the fundamental issue of, "how do we assign block ids to RGMB MGs without conflict if other MGs are going to be called before, during, or after the RGMB generation stage?"

GiudGiud replied:
So a few things:
- I m not sure you actually want different block IDs by default. We can only handle so many blocks, so the important part is to get different region IDs by default.
- Since you are using the subgenerator system, cant the block IDs be set at the time you are adding these?
- merging meshes and making sure blocks are different is not necessarily "unclean". To me it is less object oriented and more unclean to be generating a pin mesh while keeping track of the mesh (block ids) for an entire reactor.

shikhar413 replied:
So I don't think setting block IDs at the sub generator level will not be all that straightforward, since the IDs and names would need to be set together and there is no way to ensure that there isn't a conflict with later MGs. To me, it seems like temporary IDs and names will have to be set at the sub generator level and then reset at the generate level.

GiudGiud replied:
ok let's go with this then.

shikhar413 replied:
> Your best bet in the latter case might be to handle any conflicts at the point of merging. This is what libMesh does when stitching two meshes together, for instance; if one mesh has an "element 1" and the another has a different "element 1" then we don't just throw our hands up and say the meshes conflict, we renumber the latter's elements (calculating an offset to add to element ids, node ids, unique_ids) to solve the conflict.

@roystgnr can you provide more information on this comment? Does it work for mesh stitching where two MG's share the same block ID but have different names? As I implement the features requested by @GiudGiud, I am running into the issue where pin1 has let's say block ID's numbers from 10001 to 10005 and pin2 also has block IDs numbered from 10001 to 10005 but the blocks all have unique block names across the two pins. 

However, when these pins are stitched into an assembly, instead of renumbering the block IDs from 10001 to 10010 and preserving these unique block names, the block IDs are merged so that block names for pin2 are removed and the two pins continue to share the same block IDs from 10001 to 10005. I would need this functionality of remapping block IDs instead of merging them, since pin2 has no information about pin1 anymore, so an appropriate block ID offset for pin 2 cannot be known a priori. Could you please advise on if this can be done?

GiudGiud replied:
in the merge/rename routine cant you detect that they have a different block name and then nor merge the block IDs.

Or are you saying that the stitching is problematic because it thinks blocks with the same block IDs but with a different name should be merged?
In the latter case we d need to patch libmesh I think

shikhar413 replied:
So if you have a mesh block like this:

```
[Mesh]
  [pin1]
    type = PolygonConcentricCircleMeshGenerator
    polygon_size = 1.42063
    num_sides = 4
    ring_radii = '0.2 0.4'
    ring_block_ids = '10000 10001'
    ring_intervals = '1 1'
    ring_block_names = 'pin1_r0 pin1_r1'
    background_block_ids = 100002
    background_intervals = 1
    background_block_names = 'pin1_r2'
  []

  [pin2]
    type = PolygonConcentricCircleMeshGenerator
    polygon_size = 1.42063
    num_sides = 4
    ring_radii = '0.4 0.6'
    ring_block_ids = '10000 10001'
    ring_intervals = '1 1'
    ring_block_names = 'pin2_r0 pin2_r1'
    background_block_ids = 100002
    background_intervals = 1
    background_block_names = 'pin2_r2'
  []

  [assembly]
    type = CartesianIDPatternedMeshGenerator
    id_name = pin_id
    assign_type = cell
    inputs = 'pin1 pin2'
    pattern = '0 0;
               0 1;'
  []
[]
```

I was hoping to use the block name to figure out which radial level and pin ID I am currently iterating over in the merge/rename routine of `AssemblyMeshGenerator` so as to set the block names and region IDs according to what the user wants, but the block names for id's 10000 to 10002 get merged into one and so when you iterate over the pin elements in `AssemblyMeshGenerator::generate()`, at this point it's unclear whether id 10000 refers to pin1 or pin2. This block id merging leads to more complications once we start to incorporate background and duct regions in multiple assemblies as well. One possible workaround would be to tag the radial index of each pin and assembly element as an extra element integer (0+offset for radial indices in the pin, 1000+offset for assembly background and duct elements), and use this radial id, pin id, and axial_id to set the block names and region ids. What do you think?

roystgnr replied:
> Does it work for mesh stitching where two MG's share the same block ID but have different names?

The feature I described doesn't do anything at all to subdomain ids.  If we stitch two meshes and both have a subdomain 3, we assume that the final mesh should also have a subdomain 3 which is the union of the inputs, because subdomain ids don't confict in general.  I'm pointing out what we do for the values (id() and unique_id()) which do conflict, and suggesting that since in your case subdomain ids conflict too you might try an analogous tactic.

Frankly, in the case of the input file you posted, I think the correct behavior would be to die with an error.  We do require that subdomain ids and subdomain names are in 1-to-1 (or 1-to-0 when no name exists) correspondence, so we can't have two names for the same id, but the first input mesh clearly wants one name for that id and the second mesh clearly wants another.  If the input file had created automatically-chosen numeric ids rather than specifying exact numerical ids then I'd say the tactic I recommend above would apply: just choose a new number to go with the name of any subdomain id that conflicts, and renumber ... but the input file does specify exact ids and so it's not fair to the user to change them behind the user's back.

GiudGiud commented on modules/reactor/src/meshgenerators/CoreMeshGenerator.C:
```diff
@@ -51,41 +52,28 @@ CoreMeshGenerator::validParams()
-  */
-  MeshGeneratorName _reactor_params =
+  // Initialize name_id_map and current_block_id from ReactorMeshParams object stored
+  // in pin input
+  MeshGeneratorName reactor_params =
       MeshGeneratorName(getMeshProperty<std::string>("reactor_params_name", _inputs[0]));
```
seems odd / not intuitive that only the first one is queried, other reactor info stored in the other ones is ignored

shikhar413 replied:
I could add a check to make sure all values of "reactor_params_name" in `_inputs` are the same? They should all be equivalent

GiudGiud replied:
umm actually probably not. 

This reactor_params mesh generator name, it s the name of that MG at the top of the file right? 
So we could see a user define two if they're being hacky. 

let's just add a check then

shikhar413 replied:
Check has been added

GiudGiud commented on modules/reactor/src/meshgenerators/CoreMeshGenerator.C:
```diff
@@ -51,41 +52,28 @@ CoreMeshGenerator::validParams()
     mooseError("Both top_boundary_id and bottom_boundary_id must be provided in ReactorMeshParams "
                "if using extruded geometry");
-  if (!hasMeshProperty("radial_boundary_id", _reactor_params))
+  if (!hasReactorParam("radial_boundary_id"))
```
I like this API change

GiudGiud commented on modules/reactor/src/meshgenerators/CoreMeshGenerator.C:
```diff
@@ -353,82 +359,131 @@ CoreMeshGenerator::CoreMeshGenerator(const InputParameters & parameters)
+        // element is in an assembly duct or background region since it doesn't
+        // have a pin type id that matches one in the map. Infer peripheral index
+        // from pin_type
+        dof_id_type at_id = elem->get_extra_integer(atid_int);
+        unsigned int peripheral_idx = (UINT16_MAX - 1) - pt_id;
```
at_id, pt_id this is not intuitive
assembly_type_id, pin_type_id

shikhar413 replied:
Renamed

GiudGiud commented on modules/reactor/src/meshgenerators/CoreMeshGenerator.C:
```diff
@@ -353,82 +359,131 @@ CoreMeshGenerator::CoreMeshGenerator(const InputParameters & parameters)
+            (is_background_region ? _background_region_id_map[at_id][z_id]
+                                  : _duct_region_id_map[at_id][z_id][peripheral_idx - 1]);
+
+        // Swap region ids if they are different
```
is this a swap/exchange? or more like a "change"
also explain why / give more details

same on the other 'swaps'

shikhar413 replied:
These swaps occur if the block information at an axial level is different from the base 2-D information, since FancyExtruderMeshGenerator axially homogenizes the information when extruding from 2-D to 3-D. Therefore, we iterate through user defined axial variables and change them if necessary. "Change/update" is probably a better term

GiudGiud replied:
ok let's have that infomration in the comment please

shikhar413 replied:
More comments have been added to explain what is happening in these functions for PMG, AMG, and CMG

GiudGiud commented on modules/reactor/src/meshgenerators/PinMeshGenerator.C:
```diff
@@ -139,165 +133,155 @@ PinMeshGenerator::PinMeshGenerator(const InputParameters & parameters)
+      mooseError("The size of block_names must match the size of region_ids");
+    for (const auto i : make_range(_region_ids.size()))
+      if (_region_ids[i].size() != _block_names[i].size())
+        mooseError("The size of block_names must match the size of region_ids");
```
where is the check for size of region_ids being the same as the number of axial planes in extrusion

I think it's elswhere, down for a pointer

shikhar413 replied:
Doesn't look like this check exists, will add this. Thanks for the catch

shikhar413 replied:
Added a check for region_ids, background_regions_ids, and duct_region_ids

GiudGiud commented on modules/reactor/src/meshgenerators/PinMeshGenerator.C:
```diff
@@ -309,8 +293,6 @@ PinMeshGenerator::PinMeshGenerator(const InputParameters & parameters)
     for (int i = 0; i < num_sides; i++)
       boundaries_to_delete.insert(boundaries_to_delete.end(),
                                   {std::to_string(10001 + i), std::to_string(15001 + i)});
-    if (num_sides == 4)
-      boundaries_to_delete.push_back(std::to_string(20000 + _pin_type));
```
where is this done now

shikhar413 replied:
So before, we creating this boundary in PolygonConcentricCircleMeshGenerator, then rotating the pin and then assigning a new boundary to the rotated pin since the boundary wasn't propagating forward. Therefore, we had to delete the 20000 + _pin_type boundary that was created in the first step. Now, however, a lot of this logic is consolidated into a single MG using flat_side_up=true (new functionality in PCCMG), so the need to create a new boundary and delete the old 20000 + _pin_type boundary is no longer needed. We now just use the 20000 + _pin_type boundary directly

GiudGiud replied:
but you dont want to output this boundary in the final mesh right

GiudGiud replied:
Yes let's remove extra sidesets. The mesh should be as clean as possible coming out at any stage of the workflow.

Currently what stages have to have extra sidesets for future stitching?

GiudGiud commented on modules/reactor/src/meshgenerators/PinMeshGenerator.C:
```diff
@@ -139,165 +133,155 @@ PinMeshGenerator::PinMeshGenerator(const InputParameters & parameters)
+    params.set<std::vector<unsigned int>>("num_sectors_per_side") =
+        std::vector<unsigned int>(num_sides, _num_sectors);
+
+    params.set<boundary_id_type>("interface_boundary_id_shift") =
+        30000; // need to shift interface boundaries to avoid clashing
+               // with default IDs for PatternedMeshGenerator
```
is this relevant for hex pins?

shikhar413 replied:
Turns out this isn't needed for Cartesian or hex pins so I've gone ahead and removed this block

GiudGiud commented on modules/reactor/src/meshgenerators/PinMeshGenerator.C:
```diff
@@ -399,63 +380,79 @@ PinMeshGenerator::PinMeshGenerator(const InputParameters & parameters)
-    }
-  }
+  // Re-initialize name_id_map and current_block_id from ReactorMeshParams object
+  // in case variables have been modified between constructor and this method
```
in case block ids/names

GiudGiud commented:
I like the consolidation of code in pin MG between reactor types.

GiudGiud commented:
@shikhar413 I stumbled upon this piece of code  in the lowerDBlockFromSidesetGenerator.
If you make a util for it let's replace it there too

```
  // Generate a new block id if one isn't supplied.
  SubdomainID new_block_id;
  if (isParamValid("new_block_id"))
    new_block_id = getParam<SubdomainID>("new_block_id");
  else
  {
    std::set<SubdomainID> preexisting_subdomain_ids;
    mesh->subdomain_ids(preexisting_subdomain_ids);
    if (preexisting_subdomain_ids.empty())
      new_block_id = 0;
    else
    {
      const auto highest_subdomain_id =
          *std::max_element(preexisting_subdomain_ids.begin(), preexisting_subdomain_ids.end());
      mooseAssert(highest_subdomain_id < std::numeric_limits<SubdomainID>::max(),
                  "A SubdomainID with max possible value was found");
      new_block_id = highest_subdomain_id + 1;
    }
  }
```

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1084434/) on 1413f4d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21121/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21121/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 41e7a8fdd00508719d8d5aa7621bfef4c37e53f7`

shikhar413 commented:
@GiudGiud I've pushed some changes that I think should address most of the issues with RGMB block naming and region ID assignment that you addressed. The high-level strategy is now to appropriately extra element integers at the `execute` method of PinMeshGenerator and AssemblyMeshGenerator, and then use these extra element integers to "recover" the proper block name of the element at later stages if block id's get merged during pin stitching into an assembly, or assembly stitching into a core.

If block names are not provided by the user, these updates also allow you to define single block names across entire RGMB regions while using the region_id extra element integer to make material assignments, so there is now a many to one mapping between region_id and block name/ID.

The high level pseudocode at each stage is now:

##### `PinMeshGenerator` constructor:
  - set block id of each radial zone as 10,000+r, where r is the radial index (0 for inner most pin region, incrementing out to background and duct regions)
  - Set block id to 9999 if tri region is present at center pin / background region
  - Set block name to RGMB_PIN{pin_id}_R{r}, where r is the radial index described above

##### `PinMeshGenerator::generate`: 
  - Extra radial index from block name, set radial_idx extra element integer
  - Set region id from radial_idx, and axial_idx
  - Set block name as `RGMB_PIN{pin_id}_{block_name}` if user specified block names are provided, otherwise block name is `RGMB_PIN{pin_id}`

##### `AssemblyMeshGenerator` constructor:
  - set background and duct block id of each peripheral zone as 20,000+r, where r is the peripheral index (0 for inner most background region, incrementing out to duct regions)
  - Set block name to RGMB_ASSEMBLY{assembly_id}_R{r}, where r is the peripheral index described above

##### `AssemblyMeshGenerator::generate`: 
  - Set region_id according to pin_type_id, radial_idx, and axial_idx for pin elements
  - Set region_id according to peripheral_idx, and axial_idx for assembly background/duct elements
  - Remap all block id's and block names within the assembly according to block names specified by user, in case any block ids were merged during pin stitching
  - Set block name as `RGMB_ASSEMBLY{assembly_id}_{block_name}` if user specified block names are provided, otherwise block name is `RGMB_ASSEMBLY{assembly_id}`

##### `CoreMeshGenerator` constructor:
  - Nothing special

##### `AssemblyMeshGenerator::generate`: 
  - Set region_id according to pin_type_id, radial_idx, and axial_idx for pin elements
  - Set region_id according to peripheral_idx, and axial_idx for assembly background/duct elements
  - Remap all block id's and block names within the core according to block names specified by user, in case any block ids were merged during assembly stitching
  - Set block name as `RGMB_CORE_{block_name}` if user specified block names are provided, otherwise block name is `RGMB_CORE`

Also, since proper region_id assignment is crucial for this functionality, I included these as part of the unit tests for all PMG, AMG, and CMG cases (previously extra element integers were only being tested for CoreMeshGenerator)

shikhar413 commented:
As an example of how this looks now, for a Cartesian core, we have a single block (RGMB_CORE) to represent all elements, but the region IDs are still set according to the user specifications

<img width="421" alt="image" src="https://user-images.githubusercontent.com/14334316/174855740-11266458-0012-4cd6-a83c-0679ddc47a77.png">


<img width="761" alt="image" src="https://user-images.githubusercontent.com/14334316/174855818-d92c3a22-672c-4b0d-9d35-a99b395e47c4.png">

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit d4dc9d64d1a4954446c74e2f92e7f338593fcecb
+Subproject commit 8858b86dbee25b020246a4aac1c057e17b7d8cfb
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

GiudGiud commented:
Test failures isnt yours btw.

There is a MeshUtils in util/ I think. Seems like a good place to have it
nextFreeId is not descriptive enough though. nextFreeSubdomainId maybe. Maybe there's better

Ok I ll try to review again soon.
what's up with the RGMB prefix? Is it added multiple times for pins inside assmeblies inside core?

shikhar413 commented:
I include the RGMB prefix to distinguish mesh elements from non-RGMB elements so that if other elements exist the block id /name remapping doesn't happen to those elements. The RGMB prefix is applied to the last called RGMB mesh generator, i.e. all pins and assmeblies within a core will inherit the prefix RGMB_CORE(+block_name if it exists) and previous prefixes will be discarded

GiudGiud commented on framework/src/meshgenerators/PatternedMeshGenerator.C:
```diff
@@ -163,8 +170,32 @@ PatternedMeshGenerator::generate()
+  for (auto const & id_name_pair : main_subdomain_map)
+    main_subdomain_map_name_list.emplace(id_name_pair.second);
+  if (main_subdomain_map.size() != main_subdomain_map_name_list.size())
+    paramError("inputs", "The input meshes contain subdomain name maps with conflicts.");
```
needs to be more explicit. 
what is the conflict
how to fix it/ what to look for

shikhar413 replied:
Added a more useful error message to describe the subdomain name that has conflicting ids

GiudGiud commented on framework/include/meshgenerators/PatternedMeshGenerator.h:
```diff
@@ -47,6 +47,11 @@ class PatternedMeshGenerator : public MeshGenerator
   /// Holds a mesh for each row, these will be stitched together in the end
   std::vector<std::unique_ptr<ReplicatedMesh>> _row_meshes;
 
+  /// Merges the subdomain name maps between two meshes
```
Add docstring to explain what each map contains. 

Add a note about the conflict issue

GiudGiud commented on modules/reactor/include/meshgenerators/ReactorGeometryMeshBuilderBase.h:
```diff
@@ -0,0 +1,92 @@
+
+  ReactorGeometryMeshBuilderBase(const InputParameters & parameters);
+
+  /**
+   * Checks input mesh and returns max(block ID) + 1, which represents
+   * a block ID in the mesh that is not currently in use
+   * @param input mesh over which to compute next free block id
+   */
+  SubdomainID nextFreeId(MeshBase & input_mesh);
```
move this to mesh utils. getNextFreeSubdomainID

freeId is not descriptive enough

shikhar413 replied:
This is now moved to mose mesh util and the code in lowerDBlockFromSidesetGenerator also calls this function

GiudGiud commented on modules/reactor/src/meshgenerators/CoreMeshGenerator.C:
```diff
@@ -250,56 +238,76 @@ CoreMeshGenerator::CoreMeshGenerator(const InputParameters & parameters)
               std::pair<subdomain_id_type, std::vector<std::vector<subdomain_id_type>>>(
-                  assembly_type, duct_regions));
+                  assembly_type, duct_region_ids));
+
```
add a comment to say why you are updating all these maps

GiudGiud reviewed (changes requested):
getting pretty close to being ready. 

new APIs do clean things up quite a bit imo.
There's quite a bit of regolding going on. It's good to avoid that as much as possible btw. It's better to add a mostly redundant test than to regold an existing one

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonConcentricCircleMeshGeneratorBase.C:
```diff
@@ -152,6 +152,8 @@ PolygonConcentricCircleMeshGeneratorBase::validParams()
                         "Volume of concentric circles can be preserved using this function.");
   params.addParam<subdomain_id_type>("block_id_shift", 0, "Integer used to shift block IDs.");
+  params.addParam<bool>(
+      "create_interface_boundaries", true, "Whether the interface boundaries are created.");
```
are they interfaces though
seems like they are external boundaries wrt to that mesh?

GiudGiud reviewed (changes requested):
this is getting ready 
could you please clean up the commit history
notably:
- the merge commits need to go
- only regold each file once (I have not fully checked how many regolds we have here, we might be already good) and with a detailed explanation in the commit
- the "bug fix" commits need to be explicit about what bug they are fixing
- no clang format commit if it can be avoided (we have 3 here)
- any "address GiudGiud comment" should say what they actually do if it's beyond fixing spelling

shikhar413 committed:
Update PinMeshGenerator to handle block names, don't error if tri element center and intervals[0] > 1

shikhar413 committed:
Get AMG to support new block name structure

shikhar413 committed:
Get CMG to support new block name structure

shikhar413 committed:
Update documentation, refs #21120

shikhar413 committed:
CIVET fixes

shikhar413 committed:
Fix documentation

shikhar413 committed:
Increase test coverage

Conflicts:
	modules/reactor/test/tests/meshgenerators/core_mesh_generator/gold/core_hex_in.e

shikhar413 committed:
Unity build fix

shikhar413 committed:
Address @GiudGiud comments - create mergeSubdomainNameMaps and next_free_id functions, remove one-to-one mapping requirement between block name and region ID

Conflicts:
	modules/reactor/test/tests/meshgenerators/assembly_mesh_generator/gold/assembly_ductIDs_in.e
	modules/reactor/test/tests/meshgenerators/assembly_mesh_generator/gold/assembly_only_in.e
	modules/reactor/test/tests/meshgenerators/core_mesh_generator/gold/core_empty_in.e
	modules/reactor/test/tests/meshgenerators/core_mesh_generator/gold/core_hex_in.e
	modules/reactor/test/tests/meshgenerators/core_mesh_generator/gold/core_in.e
	modules/reactor/test/tests/meshgenerators/pin_mesh_generator/gold/pin_only_background_tri_intervals_in.e
	modules/reactor/test/tests/meshgenerators/pin_mesh_generator/gold/pin_only_hex_complex_in.e
	modules/reactor/test/tests/meshgenerators/pin_mesh_generator/gold/pin_only_hex_in.e
	modules/reactor/test/tests/meshgenerators/pin_mesh_generator/gold/pin_only_in.e
	modules/reactor/test/tests/meshgenerators/pin_mesh_generator/gold/pin_only_quad_intervals_in.e
	modules/reactor/test/tests/meshgenerators/pin_mesh_generator/gold/pin_only_tri_in.e
	modules/reactor/test/tests/meshgenerators/pin_mesh_generator/gold/pin_only_tri_intervals_in.e

shikhar413 committed:
Updated documentation figures, fix non unity build

Conflicts:
	large_media

shikhar413 committed:
Address @GiudGiud comments - Rename nextFreeId to getNextFreeSubdomainID and move to MooseMeshUtils

shikhar413 committed:
make interface boundaries creation optional

shikhar413 committed:
Call create_interface_boundaries from RGMB MGs

shikhar413 committed:
Switch order of if else conditions in PMGBase to avoid segfault

bug fix 2

Conflicts:
	modules/reactor/test/tests/meshgenerators/polygon_concentric_circle_mesh_generator/gold/poly_2d_full_no_interface.e

shikhar413 committed:
Add conditions to create_interface_boundaries to avoid segfault

shikhar413 committed:
Re-gold all RGMB tests to remove extra interface boundaries

shikhar413 commented:
@GiudGiud I've gone ahead and updated the commit history to the way that you requested, and also removed all Exodus file regolding from all commits except the last one. Let me know if there is anything else to be done

GiudGiud commented on modules/reactor/src/meshgenerators/AssemblyMeshGenerator.C:
```diff
@@ -314,20 +353,51 @@ AssemblyMeshGenerator::AssemblyMeshGenerator(const InputParameters & parameters)
+    }
+  }
+
+  // Remove outer pin sidesets created by PolygonConcentricCircleMeshGenerator
```
this works. An alternative, and the currently more MOOSE-way, is to delete the sidesets after stitching in the patternXXGenerator. In fact I thought the stitching would do that by default?

shikhar413 replied:
I don't believe so. In these examples these sidesets are still preserved after stitching in the patternXXGenerator

GiudGiud commented on modules/reactor/src/meshgenerators/CoreMeshGenerator.C:
```diff
@@ -249,66 +238,114 @@ CoreMeshGenerator::CoreMeshGenerator(const InputParameters & parameters)
+          continue;
+        const auto assembly_id = getMeshProperty<subdomain_id_type>("assembly_type", assembly_name);
+        const BoundaryName boundary_name = "outer_assembly_" + std::to_string(assembly_id);
+        if (std::find(boundaries_to_delete.begin(), boundaries_to_delete.end(), boundary_name) ==
```
if you change the equivalent code in AssemblyMG to a set or something else, dont forget this one

GiudGiud commented on modules/reactor/src/meshgenerators/CoreMeshGenerator.C:
```diff
@@ -249,66 +238,114 @@ CoreMeshGenerator::CoreMeshGenerator(const InputParameters & parameters)
+              std::pair<subdomain_id_type, std::vector<std::vector<subdomain_id_type>>>(
+                  pin->first, pin->second));
+        else if (pin->second != _pin_region_id_map.find(pin->first)->second)
           mooseError("Multiple region definitions for the same pin type. Check pin_type ids.\n");
```
```suggestion
          mooseError("Multiple region id definitions for the same pin type. Check pin_type ids.\n");
```

GiudGiud commented on modules/reactor/src/meshgenerators/AssemblyMeshGenerator.C:
```diff
@@ -390,115 +460,90 @@ std::unique_ptr<MeshBase>
+      // Pin type element, get region ID from pin_type, z_id, and radial_idx
+      const dof_id_type radial_idx = elem->get_extra_integer(radial_id_int);
+      const auto elem_rid = _pin_region_id_map[pin_type_id][z_id][radial_idx];
+      elem->set_extra_integer(region_id_int, elem_rid);
```
oh it's because the extrusion mesh generator did not propagate it?

GiudGiud commented on modules/reactor/src/meshgenerators/CoreMeshGenerator.C:
```diff
@@ -354,81 +391,118 @@ std::unique_ptr<MeshBase>
+  auto next_block_id = MooseMeshUtils::getNextFreeSubdomainID(*(*(_build_mesh)));
+  std::map<std::string, SubdomainID> rgmb_name_id_map;
+
+  // Loop through all mesh elements and set region ids and reassign block IDs/names
+  // if they were merged during assembly stitching
+  for (auto & elem : (*_build_mesh)->active_element_ptr_range())
```
this could be real expensive in 3D after the extrusion.

Fixing the merging of names could be done in 2D right? What cannot be done in 2D?

GiudGiud replied:
Well both the region ID and the block names depends on the user input in the pin MG, which are 2D vector axial * radial.

Save for writing a new extruder that also takes care of adding those, not sure what you can do about this?

shikhar413 replied:
Yeah, right now this step is needed to define region IDs and block names for all extruded elements

GiudGiud commented on modules/reactor/src/meshgenerators/CoreMeshGenerator.C:
```diff
@@ -354,81 +391,118 @@ std::unique_ptr<MeshBase>
-  {
-    mooseError("Expected mesh inputs to have the extra integer id: region_id.\n");
-  }
+  // Sideset 10000 does not get stitched properly when BlockDeletionGenerator
```
to what is it stitched exactly?
is this just the outer_core sideset ?

shikhar413 replied:
Typically in Reactor MG's sideset 10000 is reserved for the outer boundary of the final Reactor mesh. In the RGMB CoreMeshGenerator, however, this sideset does not get generated properly if there are multiple dummy assemblies, since we are supposed to generate independent dummy assemblies for each dummy assembly position but instead we only define one dummy assembly and populate it into multiple locations. This block of code, therefore, just copies the "outer_core" sideset (which is generated properly) into sideset 10000

GiudGiud commented on modules/reactor/src/meshgenerators/PinMeshGenerator.C:
```diff
@@ -339,29 +333,28 @@ PinMeshGenerator::PinMeshGenerator(const InputParameters & parameters)
-  declareMeshProperty("reactor_params_name", std::string(_reactor_params));
+  declareMeshProperty("pin_region_ids", region_id_map);
+  std::map<subdomain_id_type, std::vector<std::vector<std::string>>> block_name_map;
+  block_name_map[_pin_type] = _block_names;
```
how is this initialized if the user does not set it? seems to me like _block_names is an empty vector then?

shikhar413 replied:
Correct, this will be an empty vector. Later on, AMG will check whether this vector is empty to figure out whether block names were provided by the user or not

GiudGiud reviewed (changes requested):
mostly cosmetics and some discussions to have

GiudGiud commented:
Just a note, the way this checking for ID assignment conflicts is done requires all these maps to be local. So it wouldnt really be compatible with a distributed CoreMeshGenerator. It ll work well for the assembly stage which can be serial.

For memory reasons, you will want the extrusion mesh-subgenerator and the CoreMeshGenerator to run in distributed mode. Is this in scope for future works? I dont think MOOSE (cleanly) supports a switch from serial to distributed yet but it should be possible to add that from the base classes

shikhar413 commented on framework/src/utils/MooseMeshUtils.C:
```diff
@@ -291,4 +291,24 @@ isCoPlanar(const std::vector<Point> vec_pts)
 }
+
+SubdomainID
+getNextFreeSubdomainID(MeshBase & input_mesh)
```
The assumption here is that `input_mesh` is an actual mesh object where `generate()` has been called. I don't believe this would work if the mesh hasn't been created yet (e.g. before generate() is called). Which util file should this routine be placed then?

GiudGiud replied:
it could just be a class attribute in mooseMesh

shikhar413 committed:
Minor changes to documentation, code docstrings

shikhar413 commented on modules/reactor/doc/content/source/meshgenerators/CoreMeshGenerator.md:
```diff
@@ -28,7 +28,13 @@ If the core is extruded to three dimensions the top-most boundary ID must be ass
 !listing modules/reactor/test/tests/meshgenerators/core_mesh_generator/core.i block=Mesh
 
-!media reactor/meshgenerators/core_mesh_generator.png style=width:60%;
+Resulting mesh block layout:
```
Changed to "This is the resulting mesh block layout, where by default a single block is assigned to the triangular elements and another block is assigned to the quadrilateral elements:"

shikhar413 commented on modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h:
```diff
@@ -64,21 +73,21 @@ class AssemblyMeshGenerator : public MeshGenerator
-  std::map<subdomain_id_type, std::vector<std::vector<subdomain_id_type>>> _id_map;
+  std::map<subdomain_id_type, std::vector<std::vector<subdomain_id_type>>> _pin_region_id_map;
+
+  ///A mapping of block names and pin-type IDs that are used to allow for extrusion in an ReactorGeometry MeshGenerator
```
Changed to "A mapping from pin-type IDs to block names used when assigning block names during the pin stitching stage"

shikhar413 commented on modules/reactor/include/meshgenerators/ReactorMeshParams.h:
```diff
@@ -48,4 +48,11 @@ class ReactorMeshParams : public MeshGenerator
+
+  ///Tracker to assign new block ids in RGMB mesh generators.
+  ///Increments by 1 each time a new block id is requested
+  subdomain_id_type _current_block_id = 10000;
```
Correct, this has been removed

shikhar413 commented on modules/reactor/src/meshgenerators/AssemblyMeshGenerator.C:
```diff
@@ -314,20 +353,51 @@ AssemblyMeshGenerator::AssemblyMeshGenerator(const InputParameters & parameters)
+        const BoundaryName boundary_name = "outer_pin_" + std::to_string(pin_id);
+        if (std::find(boundaries_to_delete.begin(), boundaries_to_delete.end(), boundary_name) ==
+            boundaries_to_delete.end())
+          boundaries_to_delete.push_back(boundary_name);
```
Changed to "if (!std::count(boundaries_to_delete.begin(), boundaries_to_delete.end(), boundary_name))"

shikhar413 commented:
> For memory reasons, you will want the extrusion mesh-subgenerator and the CoreMeshGenerator to run in distributed mode. Is this in scope for future works? I dont think MOOSE (cleanly) supports a switch from serial to distributed in between sub-meshgenerators yet but it should be possible to add that from the base classes

Not sure I understand what you are saying completely, might be best to discuss the issue over a call some time. We have other features we would like to add to RGMB that are higher priority (adding capability to set extra element integers related to depletion IDs, adding a core periphery mesh generator to CoreMeshGenerator, etc.), but we can add this to our future workscope once the details of what to do are more clear. Would a workaround for now be to generate an exodus file for CoreMeshGenerator using the `--mesh-only` option in replicated mode and then define the mesh with FileMeshGenerator in distribtued mode for the actual run?

GiudGiud reviewed (approved):
thanks for the contribution and addressing all the reviews!

shikhar413 commented:
Thanks for the detailed review and suggestions to improve this functionality @GiudGiud
